src/protocol/boolean/solved_bits.rs

```diff
@@ -0,0 +1,265 @@
+use crate::protocol::reveal::Reveal;
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+
+#[allow(dead_code)]
+pub struct RandomBitsShare<F: Field> {
+    b_b: Vec<Replicated<F>>,
+    b_p: Replicated<F>,
+}
+
+/// This protocol tries to generate a sequence of uniformly random sharing of
+/// bits in `F_p`. Adding these 3-way secret-sharing will yield the secret
+/// `b_i ∈ {0,1}`. This protocol will abort and returns `None` if the secret
+/// number from randomly generated bits is not less than the field's prime
+/// number. Once aborted, the caller must provide a new narrowed context if
+/// they wish to call this protocol again for the same `record_id`.
+///
+/// This is an implementation of "3.1 Generating random solved BITS" from I. Damgård
+/// et al., but replaces `RAN_2` with our own PRSS implementation in lieu.
+///
+/// 3.1 Generating random solved BITS
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct SolvedBits {}
+
+impl SolvedBits {
+    // Try generating random sharing of bits, `[b]_B`, and `l`-bit long.
+    // Each bit has a 50% chance of being a 0 or 1, so there are
+    // `F::Integer::MAX - p` cases where `b` may become larger than `p`.
+    // With `Fp32BitPrime` (prime is `2^32 - 5`), that chance is around
+    // 1 * 10^-9. For Fp31, the chance of this protocol succeeding is around
+    // 1/2^3 =~ 13% (3 high bits being all 0's).
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Option<RandomBitsShare<F>>, Error> {
+        //
+        // step 1 & 2
+        //
+        let b_b = Self::generate_random_bits(ctx.clone(), record_id).await?;
+
+        //
+        // step 3, 4 & 5
+        //
+        // if b >= p, then abort by returning `None`
+        if !Self::is_less_than_p(ctx.clone(), record_id, &b_b).await? {
+            return Ok(None);
+        }
+
+        //
+        // step 6
+        //
+        // if success, then compute `[b_p]` by `Σ 2^i * [b_i]_B`
+        #[allow(clippy::cast_possible_truncation)]
+        let b_p: Replicated<F> = b_b
+            .iter()
+            .enumerate()
+            .fold(Replicated::ZERO, |acc, (i, x)| {
+                acc + &(x.clone() * F::from(2_u128.pow(i as u32)))
+            });
+
+        Ok(Some(RandomBitsShare { b_b, b_p }))
+    }
+
+    /// Generates a sequence of `l` random bit sharings in the target field `F`.
+    async fn generate_random_bits<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<Replicated<F>>, Error> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
```

Comment:
The abort rate in this protocol is too high, but here is where you can fix that.

You are looking at the number of fields in the integer, not the prime.  You can calculate the number of bits you need with:

```rust
let l = F::PRIME::BITS - F::PRIME.leading_zeros();
```

Reply:
I like this. Only need to generate 5 bits.

this logic will also help us if we decide to use the Mersenne prime 2^31 - 1 at some point in the future. Only need to generate 31 bits for that one! If we generate 32, it’ll have a 50% abort rate.